### PR TITLE
support multiple instance in cluster scope

### PR DESCRIPTION
**Pull request by msenmurugan**
* support multiple instance in cluster scope
* removal of security client
* Clone job fix
* displaying the route address fix
* update state fix
* commented an updation of PVC volume names in status
* in ui, don't create multiple instance of Black Duck in a single namespace
* removed host in default environs
* fixing for clone pod status
* fix for opssight internal hosts

pending
migration of existing Black Duck resources - another PR

### CODE REVIEW

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/472#" title="Author self-approved">msenmurugan</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [msenmurugan]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by rrati** [comment]
NM, just removed the security constraint client from apps that don't need it.

File: pkg/alert/crdhandler.go [hunk 1]
```diff
@@ -119,10 +119,28 @@ func (h *Handler) ObjectUpdated(objOld, objNew interface{}) {
 		return
 	}
 
+	var err error
+	isUpdate := false
+	if alert.GetAnnotations() == nil {
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/alert/crdhandler.go]: For some reason, it is failing with nil map. That's why added this logic

File: cmd/operator-ui/actions/blackducks.go [hunk 2]
```diff
@@ -238,6 +238,18 @@ func (v BlackducksResource) Create(c buffalo.Context) error {
...
 
+	blackducks, err := util.ListHubs(v.blackduckClient, blackduck.Spec.Namespace)
+	if err != nil {
+		return fmt.Errorf("unable to list Black Duck instances in namespace '%s' due to %+v", blackduck.Spec.Namespace, err)
+	}
+
+	// When running in cluster scope mode, custom resources do not have a namespace so the above command returns everything and we need to check Spec.Namespace.
+	for _, bd := range blackducks.Items {
+		if strings.EqualFold(bd.Spec.Namespace, blackduck.Spec.Namespace) {
+			return fmt.Errorf("due to known restriction, only one Black Duck instance per namespace is allowed. stay tuned for the updates")
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply cmd/operator-ui/actions/blackducks.go]: done

File: pkg/apps/blackduck/latest/creater.go [hunk 3]
```diff
@@ -195,26 +191,32 @@ func (hc *Creater) Ensure(blackduck *blackduckapi.Blackduck) error {
 		}
 	}
 
-	if blackduck.Spec.PersistentStorage {
-		pvcVolumeNames := map[string]string{}
-		pvcList, err := hc.kubeClient.CoreV1().PersistentVolumeClaims(blackduck.Spec.Namespace).List(metav1.ListOptions{
-			LabelSelector: fmt.Sprintf("app=%s,name=%s,component=pvc", util.BlackDuckName, blackduck.Name),
-		})
+	// Commented to verify where it is used
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/apps/blackduck/latest/creater.go]: done

File: pkg/apps/blackduck/v1/creater.go [hunk 4]
```diff
@@ -163,26 +159,32 @@ func (hc *Creater) Ensure(blackduck *blackduckapi.Blackduck) error {
 		}
 	}
 
-	if blackduck.Spec.PersistentStorage {
-		pvcVolumeNames := map[string]string{}
-		pvcList, err := hc.kubeClient.CoreV1().PersistentVolumeClaims(blackduck.Spec.Namespace).List(metav1.ListOptions{
-			LabelSelector: fmt.Sprintf("app=%s,name=%s,component=pvc", util.BlackDuckName, blackduck.Name),
-		})
+	// Commented to verify where it is used
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/apps/blackduck/v1/creater.go]: done

File: pkg/blackduck/crdhandler.go [hunk 5]
```diff
@@ -130,18 +130,40 @@ func (h *Handler) ObjectDeleted(name string) {
...
 func (h *Handler) ObjectUpdated(objOld, objNew interface{}) {
+	var err error
 	bd, ok := objNew.(*blackduckv1.Blackduck)
 	if !ok {
 		log.Error("unable to cast Black Duck object")
 		return
 	}
 
+	isUpdate := false
+	if bd.GetAnnotations() == nil {
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/blackduck/crdhandler.go]: For some reason, it is failing with nil map. That's why added this logic

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/472#" title="Author self-approved">msenmurugan</a>*, *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/472#issuecomment-503584708" title="Approved">rrati</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [msenmurugan,rrati]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/472#" title="Author self-approved">msenmurugan</a>*, *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/472#issuecomment-503584739" title="LGTM">rrati</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [msenmurugan,rrati]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**